### DISCUSSION TITLE
Unable to communicate with the robot via ROS 2

### DISCUSSION THREAD

**Discussion by tgkell** [iRobotEducation/create3_docs#84] [Questions] [answered]
I'm new to Create 3.  I set it up, updated firmware according to instructions,   I had ROS2 working from Ubuntu yesterday, but left it on the dock.  I powered it off while still on the dock and found I had to pick it up off the dock to get the lights to shut off.  This morning I put it back on the dock, it powered on and booted up, and ROS2 now shows just
$ros2 topic list
/parameter_events
/rosout

I've tried two different Ubuntu machines with the same results.  I've tried switching to CycloneDDS (re-started application on Create 3, changed environment variable RMW_IMPLEMENTATION=rmw_cyclonedds_cpp on Ubuntu), and I've tried switching it back to RMW_IMPLEMENTATION=rmw_fastrtps_cpp on Create 3 and Ubuntu, several times, with no luck)

I've tried powering off and on, and I've tried moving the bluetooth switch to "USB".

No network changes since yesterday: both Ubuntu and Create 3 are on the same network (can access the Create 3 webpage and I can ping Create 3 from Ubuntu).

I've attached the log file below, which does contain errors, although I believe the same errors were present when ROS2 was working.

If anyone has any pointers, I'd appreciate hearing them!

Cheers,
Todd
[messages.txt](https://github.com/iRobotEducation/create3_docs/files/8950877/messages.txt)

**Comment by shamlian** [comment]
This is likely a problem with your computer's network set up, or a middleware mismatch, as you suspected. Your logs look normal, to me. Can you give us more information about your network setup, which middleware you are using, and whether or not you have any custom configuration?

**Reply by justinIRBT** [reply]
One more thing to try, when you do ros2 topic list, can you instead run
ros2 topic list --no-daemon --spin-time 10
This will eliminate the possibility of a stale cache somewhere

**Reply by tgkell** [reply]
Thanks for the replies.  The setup is to the letter of the Create 3 getting started guide, and the to the letter the Setup instructions for Ubuntu here: https://iroboteducation.github.io/create3_docs/setup/ubuntu2004/.  However, I did switch it to FastDDS to get ROS2 to work.  The only thing that changed from the time it was working to the time it stopped was I powered off the Create 3, then about 8 hrs later, I docked it (powered on).

Here are some more details:
The wireless network is provided by a Netgear Nighthawk RAX200 which offers 2.4GHZ and two 5GHZ channels on the same SSID. The actual channel used is negotiated, and hmm, maybe the Create 3 doesn't like that wireless feature, but it seemed to work fine in the beginning.  I'm currently running ROS2 in an Ubuntu VM in bridged networking mode, but it was originally working on Windows WSL2 Ubuntu running in a VM.  Right now it won't work with either.

Ubuntu IP: 192.168.1.22, and can ping the Create 3
IP address of Create 3: 192.168.1.21, and the Create 3 webpage, updates etc, always work fine at that address

Since Fast DDS stopped working, I've tried Cyclone DDS and Fast DDS multiple times, being careful to save and restart application on the Create 3 each time, and on the Ubuntu side, be sure to set the RMW_IMPLEMENTATION variable correctly, exit, and start a new shell (even rebooted both sides multiple times).  There is only one Create 3, and the namespace is left blank.

The only part of my setup that I would consider non-standard is that I use bridged networking Ubuntu VMs to run ROS2.  It was working fine with Fast DDS with Windows WSL2 Ubuntu, but maybe it intermittently doesn't like WSL2, or VMware bridged networking, or maybe it doesn't like my Netgear wireless.  It was working fine but stopped after a Create 3 power cycle. Intermittent problems are the worst!

**Reply by tgkell** [reply]
Whoops, I also should have mentioned that I've tried this many times:
ros2 topic list --no-daemon --spin-time 10

**Reply by bluecamel** [reply]
> 

Maybe try increasing the spin time?  If the network is slow, it might give up early.

**Reply by shamlian** [reply]
I suspect a networking problem with the virtual machines, but it will be difficult to figure out exactly where it is. I haven't worked with VMWare in particular, but I will see what I can do to find a colleague who can help.

**Reply by tgkell** [reply]
My problem does not seem to be with the Create 3 itself.  I believe you're right that the main issue seems to be with virtual machine networking. I also suspect my wireless router with Netgear SmartConnect might be affecting things, possibly.

I installed a bare-metal (not a VM) Ubuntu machine with wired networking, and ROS2 seems to work well between it and the Create 3, and has continued working after a power cycle of the Create 3.

Furthermore, the two VMs, a Ubuntu VM, and a Windows VM running Ubuntu WSL2, both work if networking is bridged to a wired Ethernet connection on the host.  Neither VM works (anymore) with networking bridged to wireless, even with "ros2 topic list --no-daemon --spin-time 100" .

To test my wireless setup without VMs involved, I'm currently working on setting up a Raspberry Pi 3B running Ubuntu with wireless networking.  I'll post when I have results.  The 64G microSD card does boot, but it might be too big and causing weird storage-based issues, so I'm waiting for a 32G card to arrive.  Another source of weird issues might be that I've never used cloud-init before, but I should be able to get on top of that.

**Reply by shamlian** [reply]
Best of luck to you; for all of the good things it provides, I've found that virtualization breaks down when things (networking, USB) leave the host. As far as bridged networking over Wi-Fi, I have heard of a user who was having trouble with that when their computer auto-switched networks on them (for example, when there was a low signal strength for the AP they were using), where for whatever reason the VM didn't properly recover the network connection.

For what it's worth, I use a 64 GB card with a Raspberry Pi, and haven't had any trouble with that.

I'm going to leave this open since we're still in the process of trying to figure out how to make your end-to-end setup work; I'm glad at least to know that the Create is behaving itself. When you feel we've come to resolution, please make a new top-level answer and mark it as "answered."

**Reply by maddiepero** [reply]
You didn't happen to change your Domain ID number right? You might also want to run the command sudo nano ~/.bashrc and make sure you only have one export RMW_IMPLEMENTATION=rmw_fastrtps_cpp .  As far as making sure your connection gets over we found that a Bridged network connection that replicates the physical network connection and in configure adapters only the WLAN box is ticked works best. We also just tried pinging apple from the terminal in VM to test if the connection was working.

**Comment by tgkell** [comment, marked as answer]
My investigations point to VMware Fusion networking being the problem with "ros2 topic list" not working on VM machines.  Wireless is done with Netgear SmartConnect and that might be part of the problem as well.  Here are the facts:

The Create 3 docs are awesome: the team did a great job in my opinion!

Wired bare metal Ubuntu machine works
Wireless bare metal Ubuntu Pi works
VMs DO work when the Mac laptop is wired, and the VM networking is bridged explicitly to the wired interface.
VMs do NOT work when the Mac laptop is wireless only, with VM networks bridged explicitly to the wireless interface.
Fusion networking is a bit finicky on my Mac in general lately.

A big thank you to all who replied!
